### PR TITLE
bpo-29981: Update Index for set, dict, and generator 'comprehensions'

### PR DESCRIPTION
**Pull request by DahlitzFlorian**
A similar PR (#995) was made back in 2017, but was closed due to inactivity. With this PR I propose nearly the same changes.

@terryjreedy mentioned that he does not like the "all or part of" phrase. As the already existing list comprehension's glossary entry has a similar phrasing, I suggest to keep it.

I build the changes locally and the index was updated accordingly.

https://bugs.python.org/issue29981

### CODE REVIEW

File: Doc/glossary.rst [hunk 1]
```diff
@@ -304,6 +304,12 @@ Glossary
       keys can be any object with :meth:`__hash__` and :meth:`__eq__` methods.
       Called a hash in Perl.
 
+   dictionary comprehension
+      A compact way to process all or part of the elements in a sequence and
+      return a dictionary with the results. ``results = {n: n ** 2 for n in
+      range(10)}`` generates a dictionary containing key ``n`` which mapped to
```
Comment (remilapeyre) [inline Doc/glossary.rst]:
```suggestion
      range(10)}`` generates a dictionary containing key ``n`` mapped to
```

File: Doc/glossary.rst [hunk 2]
```diff
@@ -304,6 +304,12 @@ Glossary
       keys can be any object with :meth:`__hash__` and :meth:`__eq__` methods.
       Called a hash in Perl.
 
+   dictionary comprehension
+      A compact way to process all or part of the elements in a sequence and
```
Comment (remilapeyre) [inline Doc/glossary.rst]:
```suggestion
      A compact way to process all or part of the elements in a iterable and
```

File: Doc/glossary.rst [hunk 3]
```diff
@@ -1004,6 +1010,12 @@ Glossary
       reserved for rare cases where there are large numbers of instances in a
       memory-critical application.
 
+   set comprehension
+      A compact way to process all or part of the elements in a sequence and
```
Comment (remilapeyre) [inline Doc/glossary.rst]:
```suggestion
      A compact way to process all or part of the elements in a iterable and
```

File: Doc/glossary.rst [hunk 4]
```diff
@@ -1004,6 +1010,12 @@ Glossary
...
+      return a set with the results. ``results = {c for c in 'abracadabra' if
+      c not in 'abc'}`` generates the set of strings with ``{'r', 'd'}``.  See
```
Comment (remilapeyre) [inline Doc/glossary.rst]:
```suggestion
      c not in 'abc'}`` generates the set of strings ``{'r', 'd'}``.  See
```

File: Doc/library/stdtypes.rst [hunk 5]
```diff
@@ -4114,6 +4114,12 @@ The constructors for both classes work the same:
 .. class:: set([iterable])
            frozenset([iterable])
 
+   Sets can be created by several means:
+
+   * Use a comma-separated list of elements within braces: ``{'jack', 'sjoerd'}``
```
Comment (remilapeyre) [inline Doc/library/stdtypes.rst]:
```suggestion
   * Using a comma-separated list of elements within braces: ``{'jack', 'sjoerd'}``
```

File: Doc/library/stdtypes.rst [hunk 6]
```diff
@@ -4114,6 +4114,12 @@ The constructors for both classes work the same:
...
+   * Use a set comprehension: ``{c for c in 'abracadabra' if c not in 'abc'}``
```
Comment (remilapeyre) [inline Doc/library/stdtypes.rst]:
```suggestion
   * Using a set comprehension: ``{c for c in 'abracadabra' if c not in 'abc'}``
```

File: Doc/library/stdtypes.rst [hunk 7]
```diff
@@ -4114,6 +4114,12 @@ The constructors for both classes work the same:
...
+   * Use the type constructor: ``set()``, ``set('foobar')``, ``set(['a', 'b', 'foo'])``
```
Comment (remilapeyre) [inline Doc/library/stdtypes.rst]:
```suggestion
   * Using the type constructor: ``set()``, ``set('foobar')``, ``set(['a', 'b', 'foo'])``
```

File: Doc/library/stdtypes.rst (hunk 5 above)
Comment (remilapeyre) [reply Doc/library/stdtypes.rst]: Maybe a reminder that `{}` is the empty dictionary could be helpful here for novice users.

File: Doc/library/stdtypes.rst [hunk 8]
```diff
@@ -4309,6 +4315,14 @@ pairs within braces, for example: ``{'jack': 4098, 'sjoerd': 4127}`` or ``{4098:
            dict(mapping, **kwarg)
            dict(iterable, **kwarg)
 
+   Dictionaries can be created by several means:
+
+   * Use a comma-separated list of ``key: value`` pairs within braces:
```
Comment (remilapeyre) [inline Doc/library/stdtypes.rst:4323]:
```suggestion
   * Using a comma-separated list of ``key: value`` pairs within braces:
```

File: Doc/library/stdtypes.rst [hunk 9]
```diff
@@ -4309,6 +4315,14 @@ pairs within braces, for example: ``{'jack': 4098, 'sjoerd': 4127}`` or ``{4098:
...
+     ``{'jack': 4098, 'sjoerd': 4127}`` or ``{4098: 'jack', 4127: 'sjoerd'}``
+   * Use a dict comprehension: ``{}``, ``{x: x ** 2 for x in range(10)}``
```
Comment (remilapeyre) [inline Doc/library/stdtypes.rst:4325]: Use -> Using

Also, could we make `{}` part of the first item? It's the most common way to build a dict so it should be more pre-eminent, it's also not really a dictionary comprehension.

File: Doc/library/stdtypes.rst [hunk 10]
```diff
@@ -4309,6 +4315,14 @@ pairs within braces, for example: ``{'jack': 4098, 'sjoerd': 4127}`` or ``{4098:
...
+   * Use the type constructor: ``dict()``,
```
Comment (remilapeyre) [inline Doc/library/stdtypes.rst:4326]:
```suggestion
   * Using the type constructor: ``dict()``,
```

**Review by remilapeyre — COMMENTED** [review]
Thanks for improving the documentation @DahlitzFlorian !

File: Doc/library/stdtypes.rst (hunk 9 above)
Comment (DahlitzFlorian) [reply Doc/library/stdtypes.rst:4325]: I would not change Use to Using as Use is the imperative form and highly preferred. See also https://github.com/python/cpython/pull/995/files#r120774265

**Comment by DahlitzFlorian** [comment]
Thanks @remilapeyre for the review! I applied some of your suggestions. However, I would not change Use to Using as Use is the imperative form and highly preferred. See also: https://github.com/python/cpython/pull/995/files#r120774265

File: Doc/glossary.rst [hunk 11]
```diff
@@ -304,6 +304,12 @@ Glossary
       keys can be any object with :meth:`__hash__` and :meth:`__eq__` methods.
       Called a hash in Perl.
 
+   dictionary comprehension
+      A compact way to process all or part of the elements in a iterable and
```
Comment (akuchling) [inline Doc/glossary.rst]: Grammar/typo: it should be 'an iterable'.

File: Doc/glossary.rst [hunk 12]
```diff
@@ -1004,6 +1010,12 @@ Glossary
       reserved for rare cases where there are large numbers of instances in a
       memory-critical application.
 
+   set comprehension
+      A compact way to process all or part of the elements in a iterable and
```
Comment (akuchling) [inline Doc/glossary.rst]: 'An iterable', again.

**Review by akuchling — CHANGES_REQUESTED** [review]
The text reads well. I only noticed two minor grammar corrections.  Thanks!

**Comment by DahlitzFlorian** [comment]
Thanks for the review @akuchling, I applied the requested changes.

**Review by akuchling — APPROVED** [review]

**Comment by DahlitzFlorian** [comment]
Thanks for accepting the PR @akuchling! Is there anything left to do or can it be merged?

**Comment by akuchling** [comment]
@DahlitzFlorian: on re-reviewing this, can you please move the 'Sets/Dicts can be created...' paragraphs in `stdtypes.rst` so they come *after* the paragraph that begins "Return a set/dict"?  I think it reads oddly to have the second paragraph begin "Return <thing>".

**Comment by DahlitzFlorian** [comment]
@akuchling changed it accordingly, thanks for the hint!

**Comment by terryjreedy** [comment]
Without looking at the resulting html, the additions look good to me.  I think this should have a news entry such as:
"Improve documentation of sets and dicts, including new glossary entries for dictionary and set comprehensions."

**Comment by akuchling** [comment]
I think most documentation changes don't need a news entry, unless they're doing something large like adding a section or a whole new document. "Improved docs on topic X" doesn't tell you very much, and users don't need to modify code to accommodate doc changes.

**Comment by terryjreedy** [comment]
I have a lower threshhold -- is something added that users might want to read.  But whoever does the work to push a PR over the finish line gets to decide.

**Comment by DahlitzFlorian** [comment]
Thanks for the guidance @remilapeyre and @akuchling 🎉

**Pull request merged**